src/cmap/connection_requester.rs

```diff
@@ -0,0 +1,113 @@
+use tokio::sync::{mpsc, oneshot};
+
+use super::{worker::PoolWorkerHandle, Connection};
+use crate::{
+    error::{ErrorKind, Result},
+    options::StreamAddress,
+    runtime::AsyncJoinHandle,
+    RUNTIME,
+};
+use std::time::Duration;
+
+/// Handle for requesting Connections from the pool.
+#[derive(Clone, Debug)]
+pub(super) struct ConnectionRequester {
+    address: StreamAddress,
+    sender: mpsc::UnboundedSender<oneshot::Sender<RequestedConnection>>,
+    handle: PoolWorkerHandle,
+}
+
+impl ConnectionRequester {
+    /// Returns a new requester/receiver pair.
+    pub(super) fn new(
+        address: StreamAddress,
+        handle: PoolWorkerHandle,
+    ) -> (Self, ConnectionRequestReceiver) {
+        let (sender, receiver) = mpsc::unbounded_channel();
+        (
+            Self {
+                address,
+                sender,
+                handle,
+            },
+            ConnectionRequestReceiver { receiver },
+        )
+    }
+
+    /// Request a connection from the pool that owns the receiver end of this requester.
+    /// Returns an error if it takes longer than wait_queue_timeout before either a connection is
+    /// received or an establishment begins.
+    pub(super) async fn request(&self, wait_queue_timeout: Option<Duration>) -> Result<Connection> {
+        let (sender, receiver) = oneshot::channel();
+
+        // this only errors if the receiver end is dropped, which can't happen because
+        // we own a handle to the worker, keeping it alive.
+        self.sender.send(sender).unwrap();
+
+        // similarly, the receiver only returns an error if the sender is dropped, which
```

Comment:
Could you move this comment to the call to `receiver.await.unwrap()`? It confused me at first because I thought it was referring to the timeout error